src/store/features/blocks/blocks.ts

```diff
@@ -0,0 +1,97 @@
+import tbConf, {ToolboxConfiguration} from "../../../toolboxconfig"
+import {produce} from 'immer'
+import {MutableRefObject} from "react";
+
+// Define actions
+const TOGGLE_CATEGORY = 'category/toggle'
+const TOGGLE_BLOCK_DESCRIPTION = 'block/toggleDescription'
+
+// Type actions
+type ToggleCategory = {
+  type: typeof TOGGLE_CATEGORY,
+  payload: {
+    category: number,
+    blocksDivRef: MutableRefObject<any>
+  }
+}
+
+type ToggleBlockDescription = {
+  type: typeof TOGGLE_BLOCK_DESCRIPTION,
+  payload: {
+    block: number,
+    blockRef: MutableRefObject<any>
+  }
+}
+
+// Action creators
+export const toggleCategoryAction = (categoryId: number, categoryRef: MutableRefObject<any>): ToggleCategory => ({
+  type: TOGGLE_CATEGORY,
+  payload: {
+    category: categoryId,
+    blocksDivRef: categoryRef
+  }
+})
+
+export const toggleBlockDescriptionAction = (blockId: number, blockRef: MutableRefObject<any>): ToggleBlockDescription => ({
+  type: TOGGLE_BLOCK_DESCRIPTION,
+  payload: {
+    block: blockId,
+    blockRef: blockRef
+  }
+})
+
+// Reducers
+type Actions = ToggleCategory | ToggleBlockDescription
+
+let calcHeight = true
+const blocksReducer = (state: ToolboxConfiguration = tbConf, action: Actions) => {
+  switch (action.type) {
+    case 'category/toggle':
+      // Calculate total height of all child block components for accordion animation
+      let maxHeight = 0;
+      (action.payload.blocksDivRef.current.childNodes as NodeListOf<HTMLDivElement>).forEach((block: HTMLDivElement) => {
+        maxHeight += block.clientHeight
+      })
+      // Update state
+      return produce(state, draftState => {
+        const foundCategory = draftState.categories.find(c => c.id === action.payload.category)
+        if (foundCategory) {
+          foundCategory.toggled = !foundCategory.toggled
+          foundCategory.toggled ? foundCategory.maxHeight = maxHeight : foundCategory.maxHeight = 0
+          // Set calcHeight
+          if (foundCategory.openDesc) {
+            calcHeight = false
+          } else {
+            calcHeight = true
+          }
+        }
+      })
+    case 'block/toggleDescription':
+      return produce(state, draftState => {
+        const parentCategory = draftState.categories.find(c => c.blocks.find(b => b.id === action.payload.block))
+        const addHeight = action.payload.blockRef.current.getBoundingClientRect().height
+        if (parentCategory) {
+          parentCategory.openDesc = action.payload.block
+          parentCategory.blocks.forEach(b => {
+            if (b.id === action.payload.block) {
+              if (b.toggled) {
+                parentCategory.openDesc = undefined
```

Comment:
Prefer using `null` in this case. If the variable is meant to not have an object, it should be null. Undefined is only a temporary state when we haven't defined a value for this variable yet. Null is to say the variable exists and we want it to have no value.

Reply:
Roger that, thanks for taking the time to explain. Fixed and will keep in mind for the future.